### PR TITLE
Better Sharding for dsv3 moe layer

### PR DESCRIPTION
**Pull request by suexu1025**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -300,8 +300,13 @@ def __init__(
     self.quant = quant
     self.rngs = rngs
 
-    self.wi_kernel_axes = ("exp", "embed_no_exp", "mlp")
-    self.wo_kernel_axes = ("exp", "mlp", "embed_no_exp")
+    # special sharding for dsv3
```
Comment (gobbleturk) [inline src/MaxText/layers/moe.py]: @richjames0 for all sharding changes =D

File: src/MaxText/layers/moe.py [hunk 2]
```diff
@@ -300,8 +300,13 @@ def __init__(
...
+    if self.config.num_experts == 256:
```
Comment (gobbleturk) [inline src/MaxText/layers/moe.py]: We need a better logic for such a specific conditional. It should be controllable via a base.yml config argument like ("expert_first_dim")

This change doesn't surprise me that expert first is the most performant. Are there are any downsides to just flipping it here? Other than having to modify all of our checkpoint conversion scripts =D

Comment (gobbleturk) [reply src/MaxText/layers/moe.py]: If modifying the checkpoint conversion scripts is really the hardest part we can support both options (with a base.yml config) for a while with a deprecation warning on the current behavior to migrate to the new one.

**Review by gobbleturk — COMMENTED** [review]
@richjames0 for thoughts =D

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -900,9 +905,15 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
 
     # w0, w1, wo needs to be un sharded on fsdp / fsdp_transpose axis, so use
     # mlp_no_fsdp axis
-    w0_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    w1_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    wo_pspec = nn.logical_to_mesh_axes(("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+    # special sharding for dsv3 to remove overhead between gmm/AG
+    if self.config.num_experts == 256:
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Similar comment with Matt. For instance, we could have a flag similar to `expert_shard_attention_option`, and we could have `expert_shard_mlp_option` or similar. 

This condition `self.config.num_experts == 256` may raise up question, how about 128 or 512 experts?

**Review by RissyRan — COMMENTED** [review]
Thanks Qinwen! Could you also attach test results in the description?

File: src/MaxText/layers/moe.py (hunk 2 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: I added `shard_by_num_of_expert` to only enable it when it expert_dim dividable by fdsp_shard_dim

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: I added shard_by_num_of_expert to only enable it when it expert_dim dividable by fdsp_shard_dim

**Comment by suexu1025** [comment]
> Thanks Qinwen! Could you also attach test results in the description?
pls see the test results here
https://screenshot.googleplex.com/5xAhsHPUdHFS6zc

File: src/MaxText/layers/moe.py [hunk 4]
```diff
@@ -427,6 +432,9 @@ def get_topk(self, gate_logits, pre_bias_logits, rngs=None):
 
     return top_k_weights, top_k_indices
 
+  def shard_by_num_of_expert(self) -> bool:
```
Comment (gobbleturk) [inline src/MaxText/layers/moe.py]: I would remove this - a user would not expect that setting "fsdp_shard_on_exp" is not enough. If you want to perform these checks I will perform them in pyconfig and fail loudly when they are not met when fsdp_shard_on_exp is true

File: src/MaxText/layers/moe.py [hunk 5]
```diff
@@ -300,8 +300,13 @@ def __init__(
     self.quant = quant
     self.rngs = rngs
 
-    self.wi_kernel_axes = ("exp", "embed_no_exp", "mlp")
-    self.wo_kernel_axes = ("exp", "mlp", "embed_no_exp")
+    # special sharding for dsv3
+    if self.shard_by_num_of_expert():
+      self.wi_kernel_axes = ("embed_no_exp", "exp", "mlp")
```
Comment (gobbleturk) [inline src/MaxText/layers/moe.py]: Should this actually be ("embed_no_exp", None, "mlp")? We never actually want the expert physical axis to shard this dimension

Comment (gobbleturk) [reply src/MaxText/layers/moe.py]: Similar for below, I think "exp" should be replaced by None in all of these? This is identical to what you have when exp=1 which we can enforce in pyconfig

File: src/MaxText/layers/moe.py (hunk 4 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: I changed the validation to pyconfig

**Comment by RissyRan** [comment]
> > Thanks Qinwen! Could you also attach test results in the description?
> > pls see the test results here
> > https://screenshot.googleplex.com/5xAhsHPUdHFS6zc

Thanks! I meant to say the original links, so that we could directly to see profiles or anything. One example [here](https://github.com/AI-Hypercomputer/maxtext/pull/2370), usually put original links for reviewers.

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Could you move `# special sharding for dsv3` inside of `if self.config.fsdp_shard_on_exp:` branch?

File: src/MaxText/layers/moe.py [hunk 6]
```diff
@@ -900,9 +906,15 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
 
     # w0, w1, wo needs to be un sharded on fsdp / fsdp_transpose axis, so use
     # mlp_no_fsdp axis
-    w0_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    w1_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    wo_pspec = nn.logical_to_mesh_axes(("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+    # special sharding for dsv3 to remove overhead between gmm/AG
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Same. Could you move this comment inside of `if self.config.fsdp_shard_on_exp:`?

File: src/MaxText/pyconfig.py [hunk 7]
```diff
@@ -1049,6 +1050,11 @@ def validate_ring_of_experts_parallelism(raw_keys):
   if raw_keys["use_ring_of_experts"] and not using_expert_parallelism(raw_keys):
     raise ValueError("Ring-of-experts requires expert-parallelism to be enabled.")
 
+def validate_shard_fsdp_on_expert_parallelism(raw_keys):
+  if raw_keys["fsdp_shard_on_exp"] and raw_keys["num_experts"] % raw_keys["ici_fsdp_parallelism"]!=0:
```
Comment (RissyRan) [inline src/MaxText/pyconfig.py:1054]: Could you help add one more assertion for TP sharding? if `raw_keys["fsdp_shard_on_exp"] and (using_tensor_parallelism(raw_keys) or using_expert_parallelism(raw_keys) --> raise only supported fsdp in this config`

File: src/MaxText/layers/moe.py (hunk 6 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: updated

File: src/MaxText/pyconfig.py (hunk 7 above)
Comment (suexu1025) [reply src/MaxText/pyconfig.py:1054]: thank you, added

File: src/MaxText/pyconfig.py [hunk 8]
```diff
@@ -1049,6 +1050,11 @@ def validate_ring_of_experts_parallelism(raw_keys):
   if raw_keys["use_ring_of_experts"] and not using_expert_parallelism(raw_keys):
     raise ValueError("Ring-of-experts requires expert-parallelism to be enabled.")
 
+def validate_shard_fsdp_on_expert_parallelism(raw_keys):
+  if raw_keys["fsdp_shard_on_exp"] and raw_keys["num_experts"] % raw_keys["ici_fsdp_parallelism"]!=0: 
+    raise ValueError("fsdp_shard_on_exp requires num_experts is divisiable by ici_fsdp_parallelism.")
+  if raw_keys["fsdp_shard_on_exp"] and (using_tensor_parallelism(raw_keys) or useing_expert_parallelism(raw_keys)): 
+    raise ValueError("fsdp_shard_on_exp requires ici_expert_parallelism = 1 and ici_tensor_parallelism/ici_tensor_transpose_parallelism = 1.")
```
Comment (RissyRan) [inline src/MaxText/pyconfig.py:1057]: Nit: probably just say `fsdp_shard_on_exp does not support EP and TP shardings`?

**Review by RissyRan — APPROVED** [review]
Thanks @suexu1025

File: src/MaxText/layers/attention_mla.py [hunk 9]
```diff
@@ -398,7 +398,7 @@ def _init_projections(self, inputs_q_shape: Tuple, inputs_kv_shape: Tuple) -> No
           out_features_shape=self.q_lora_rank,
           axis=-1,
           kernel_init=self.kernel_init,
-          kernel_axes=("embed", "q_lora"),
+          kernel_axes=("embed", "q_lora_up_proj"),
```
Comment (RissyRan) [inline src/MaxText/layers/attention_mla.py:401]: This part is helpful for all MLA models or only DS V3? 

cc @richjames0 another case.
